### PR TITLE
[#1] Add watch task

### LINKED ISSUE #1: Bundle the react-native packager
Thank you for the great work and efforts that makes the complex `re-natal` into a simple handy tool.

Now there's a separate project for React Native packager.
https://github.com/facebook/metro

It will be more convenient if `clj-rn` can start packager server and simplify basic dev workflow.

The desired workflow may like:
```
$ npx react-native install    # only once if there's no native code change
$ clj -R repl -m clj-rn.main watch
```

It has to be clarify how deployment of the app itself is done as it is sometimes necessary to update it (and it's at least needed once).

Specifically both `react-native start` and `react-native run-*` should be replaced and made irrelevant.

Some links:

* https://github.com/facebook/react-native/tree/master/local-cli/runAndroid
* https://github.com/facebook/react-native/tree/master/local-cli/runIOS
* https://github.com/facebook/react-native/blob/master/local-cli/cliEntry.js

### PR DESCRIPTION
**Pull request by endenwer**
fixes #1

### Summary:
Add watch task. This task starts react native packager, figwheel, repl and then run build. Also add new configs `:figwheel-options` and `:builds`.

### Review notes:
Example project https://github.com/endenwer/clj-rn-example. 
Commit https://github.com/endenwer/clj-rn-example/commit/4009e6fc2797f571f6340f7249b1312074effba1 shows what should be done to setup re-natal project to use clj-rn.

### Testing notes (optional):
Clone example repo and follow "usage" section.

status: ready

### CODE REVIEW

File: deps.edn [hunk 1]
```diff
@@ -1,2 +1,4 @@
-{:deps  {org.clojure/tools.cli {:mvn/version "0.3.7"}}
- :paths ["src" "resources"]}
+{:paths ["src" "resources"]
+ :deps  {org.clojure/tools.cli {:mvn/version "0.3.7"}
+         figwheel-sidecar      {:mvn/version "0.5.16-SNAPSHOT"
```
Comment (jeluard) [inline deps.edn]: Any reason to use a snapshot?

File: src/clj_rn/shell.clj [hunk 2]
```diff
@@ -1,6 +1,16 @@
-(ns clj-rn.shell)
+(ns clj-rn.shell
+  (:require [clojure.string :as str]
+            [clojure.java.io :as io]))
+(import 'java.lang.Runtime)
```
Comment (jeluard) [inline src/clj_rn/shell.clj]: Is this expected to be out of the ns declaration?

**Comment by jeluard** [comment]
@endenwer Impressive work! Would it make sense to improve the main readme to reflect this? Could incorporate some of what you did to migrate `re-natal` projects?

File: deps.edn (hunk 1 above)
Comment (endenwer) [reply deps.edn]: Yes. This is required for new version of clojurescript. I did the same here https://github.com/status-im/status-react/pull/4064/files

File: src/clj_rn/shell.clj (hunk 2 above)
Comment (endenwer) [reply src/clj_rn/shell.clj]: No reason. I will change it.

File: deps.edn (hunk 1 above)
Comment (jeluard) [reply deps.edn]: Non snapshot version is out: https://clojars.org/figwheel-sidecar/versions/0.5.16

**Comment by endenwer** [comment]
I can add instruction how to setup it for re-natal project. And also add example of config file.

File: deps.edn (hunk 1 above)
Comment (endenwer) [reply deps.edn]: I will fix it.

**Comment by jeluard** [comment]
@endenwer Also would you be interested in migrating status codebase to this model if there is a bounty for it?

**Comment by endenwer** [comment]
Yeah. I can do it. But there is one problem. Status also use `hawk` and `re-frisk`. For now I have only one idea on how to do it: add ability to call tasks from `clj-rn` directly from clj code and then change `build.clj` in status to use this tasks and also to start `hawk` and `re-frisk`.

**Comment by endenwer** [comment]
Done.

File: clj-rn.conf.example.edn [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                          :optimizations :none}}
+          {:id           :android
+           :source-paths ["src" "env/dev"]
+           :figwheel     true
+           :compiler     {:output-to     "target/android/app.js"
+                          :main          "env.android.main"
+                          :output-dir    "target/android"
+                          :optimizations :none}}]
+
+ :js-modules ["realm"]
```
Comment (dmitryn) [inline clj-rn.conf.example.edn:19]: Haven't found `js-modules` option in README. Not clear what is this option about and how to use it. Maybe it makes sense to add more information about possible config options to README.

Comment (endenwer) [reply clj-rn.conf.example.edn:19]: updated readme.

File: src/clj_rn/main.clj [hunk 4]
```diff
@@ -67,57 +97,98 @@
...
+    (core/enable-source-maps)
+    (core/write-env-dev hosts-map)
+    (rebuild-index-files build-ids hosts-map)
+    (execute-react-native-cli "start" true)
+    (when start-figwheel
+      (ra/start-figwheel!
+       {:build-ids build-ids
+        :all-builds builds
+        :figwheel-options figwheel-options}))
+    (when start-app (run-builds build-ids))
```
Comment (dmitryn) [inline src/clj_rn/main.clj]: As running `watch` task right now, it does so many things (some of them could take a while) that it's hard to follow on which step i'm currently on and what's happening at the moment. Maybe we could print step name before/after each task like:
```
Starting the app..
...
Starting cljs repl...
...
```
in different font & color.

**Review by dmitryn — APPROVED** [review]
Works fine for me in example app https://github.com/endenwer/clj-rn-example
p.s. i had to change `sha` in `deps.edn` to https://github.com/endenwer/clj-rn/commit/91638d571e659e83c37282b1ad9b779be3a49fc0 for example app to properly run

File: src/clj_rn/main.clj (hunk 4 above)
Comment (endenwer) [reply src/clj_rn/main.clj]: `react-native run-*` takes most of the space. Without it output looks more distinct:
```
Patching file node_modules/metro/src/Server/index.js to serve *.map files.
Source maps enabled.
index.ios.js was regenerated
Host in RCTWebSocketExecutor.m was updated
Dev server host for iOS: localhost
Figwheel: Starting server at http://0.0.0.0:3449
Figwheel: Watching build - ios
Figwheel: Cleaning build - ios
Compiling build :ios to "target/ios/app.js" from ["src" "env/dev"]...
Scanning folders for symlinks in /Users/endenwer/projects/clj-rn-example/node_modules (19ms)
┌──────────────────────────────────────────────────────────────────────────────┐
│                                                                              │
│  Running Metro Bundler on port 8081.                                         │
│                                                                              │
│  Keep Metro running while developing on any JS projects. Feel free to        │
│  close this tab and run your own Metro instance if you prefer.               │
│                                                                              │
│  https://github.com/facebook/react-native                                    │
│                                                                              │
└──────────────────────────────────────────────────────────────────────────────┘

Looking for JS files in
   /Users/endenwer/projects/clj-rn-example


Metro Bundler ready.

Loading dependency graph, done.
Successfully compiled build :ios to "target/ios/app.js" in 30.254 seconds.
Figwheel: Starting nREPL server on port: 7888
WARNING: boolean? already refers to: #'clojure.core/boolean? in namespace: mranderson047.toolsanalyzerjvm.v0v6v9.toolsanalyzer.v0v6v7.clojure.tools.analyzer.utils, being replaced by: #'mranderson047.toolsanalyzerjvm.v0v6v9.toolsanalyzer.v0v6v7.clojure.tools.analyzer.utils/boolean?
WARNING: boolean? already refers to: #'clojure.core/boolean? in namespace: mranderson047.toolsanalyzerjvm.v0v6v9.toolsanalyzer.v0v6v7.clojure.tools.analyzer, being replaced by: #'mranderson047.toolsanalyzerjvm.v0v6v9.toolsanalyzer.v0v6v7.clojure.tools.analyzer.utils/boolean?
Launching ClojureScript REPL for build: ios
Figwheel Controls:
          (stop-autobuild)                ;; stops Figwheel autobuilder
          (start-autobuild id ...)        ;; starts autobuilder focused on optional ids
          (switch-to-build id ...)        ;; switches autobuilder to different build
          (reset-autobuild)               ;; stops, cleans, and starts autobuilder
          (reload-config)                 ;; reloads build config and resets autobuild
          (build-once id ...)             ;; builds source one time
          (clean-builds id ..)            ;; deletes compiled cljs target files
          (print-config id ...)           ;; prints out build configurations
          (fig-status)                    ;; displays current state of system
          (figwheel.client/set-autoload false)    ;; will turn autoloading off
          (figwheel.client/set-repl-pprint false) ;; will turn pretty printing off
  Switch REPL build focus:
          :cljs/quit                      ;; allows you to switch REPL to another build
    Docs: (doc function-name-here)
    Exit: :cljs/quit
 Results: Stored in vars *1, *2, *3, *e holds last exception object
Prompt will show when Figwheel connects to your application
```
react native packager runs in parallel and figwheel already has good output. I thought to move `react-native run-*` output to log file and if it fails just tell user to see the log in the file. Maybe create issue for this.

**Comment by jeluard** [comment]
@endenwer Regarding status: how about having a local build.clj that would call the relevant `clj-rn` logic ?

**Comment by endenwer** [comment]
Yes. That's what i suggested. But I need to move code from `watch` task into function so it can be called from outside. Not a big change, I will do it here then.

**Comment by jeluard** [comment]
@endenwer Perfect!

**Comment by jeluard** [comment]
@endenwer Oh and I created a bounty for this: https://github.com/status-im/status-react/issues/5226

**Comment by endenwer** [comment]
@jeluard Awesome. I finished this issue.

**Comment by endenwer** [comment]
I updated example repo https://github.com/endenwer/clj-rn-example deps.edn so everyone can test it.

**Comment by jeluard** [comment]
@endenwer Got this message while trying to test with avd: `[Failed to validate "-i avd": Must be "simulator", or "real"]`

**Comment by endenwer** [comment]
@jeluard  You need to use `-a` for android. Like this `clj -R:dev -m clj-rn.main watch -p android -a avd`

**Comment by jeluard** [comment]
With your latest changes `run-` is launched by default every time `clj` is run? What if `react-native run-*` throws an error? Looks like I am running into this.

**Comment by endenwer** [comment]
Yes, `run-` is launched by default. If it throws an error you can restart task or start `react-native run-*` separately.

**Comment by jeluard** [comment]
Right now there doesn't appear to be any error message if `run-` fails. Is that expected? Or maybe it's not started somehow in my case?

**Comment by endenwer** [comment]
`react-native run-` shows error like this`** BUILD FAILED **`. It is shown before cljs repl output.

**Comment by jeluard** [comment]
On linux I get the following: `react-native is not found. Please install it first.`
But 
```
❯ which react-native 
/home/julien/.npm-global/bin/react-native
```

**Comment by endenwer** [comment]
It uses local installation. This is strange what you don't have it installed locally because it is in `package.json`. Maybe you didn't run `npm install`.

**Comment by jeluard** [comment]
@endenwer I have it installed as the `which` command shows.

**Comment by endenwer** [comment]
I can add ability to use globally installed react-native. First check if it is installed locally then check if it is installed globally. But it is not recommend to use globally installed react-native because it can have incompatible version.

**Comment by jeluard** [comment]
Good point. Maybe the error message could be improved so that it is less confusing?
Like suggesting to run `yarn` if `react-native` is not found and detailing the global one won't be used (if it exists)? WDYT?

**Comment by endenwer** [comment]
Maybe like this `react-native is not installed locally. If you don't have it in "package.json" run "npm install react-native --save". Note what globally installed react-native won't be used.`

**Comment by jeluard** [comment]
Sounds good!

**Comment by endenwer** [comment]
I changed the message and updated example repo.

**Comment by jeluard** [comment]
Thanks for your contribution @endenwer !

**Comment by andytudhope** [comment]
Hey @endenwer - this bounty should have been confirmed and paid out a month ago, and I am sincerely sorry for the delay. We have been having some issues with the platform, but I have taken over again and am straightening them all out. It has now been paid - thank you so much for your contribution to this awesome new library and keep em coming - we ❤️ having you around here!!
